src/util_libc.rs

```diff
@@ -76,29 +81,49 @@ pub fn sys_fill_exact(
 
 // A "weak" binding to a C function that may or may not be present at runtime.
 // Used for supporting newer OS features while still building on older systems.
-// F must be a function pointer of type `unsafe extern "C" fn`. Based off of the
-// weak! macro in libstd.
+// Based off of the DlsymWeak struct in libstd:
+// https://github.com/rust-lang/rust/blob/1.61.0/library/std/src/sys/unix/weak.rs#L84
+// except that the caller must manually cast self.ptr() to a function pointer.
 pub struct Weak {
     name: &'static str,
-    addr: LazyUsize,
+    addr: AtomicPtr<c_void>,
 }
 
 impl Weak {
+    // A non-null pointer value which indicates we are uninitialized. This
+    // constant should ideally not be a valid address of a function pointer.
+    // However, if by chance libc::dlsym does return UNINIT, there will not
+    // be undefined behavior. libc::dlsym will just be called each time ptr()
+    // is called. This would be inefficient, but correct.
+    // TODO: Replace with core::ptr::invalid_mut(1) when that is stable.
+    const UNINIT: *mut c_void = 1 as *mut c_void;
+
     // Construct a binding to a C function with a given name. This function is
     // unsafe because `name` _must_ be null terminated.
     pub const unsafe fn new(name: &'static str) -> Self {
         Self {
             name,
-            addr: LazyUsize::new(),
+            addr: AtomicPtr::new(Self::UNINIT),
         }
     }
 
-    // Return a function pointer if present at runtime. Otherwise, return null.
-    pub fn ptr(&self) -> Option<NonNull<libc::c_void>> {
-        let addr = self.addr.unsync_init(|| unsafe {
-            libc::dlsym(libc::RTLD_DEFAULT, self.name.as_ptr() as *const _) as usize
-        });
-        NonNull::new(addr as *mut _)
+    // Return the address of a function if present at runtime. Otherwise,
+    // return null. Multiple callers can call ptr() concurrently. It will
+    // always return _some_ value returned by libc::dlsym. However, the
+    // dlsym function may be called multiple times.
+    pub fn ptr(&self) -> Option<NonNull<c_void>> {
+        // Despite having only a single atomic variable (self.addr), we still
+        // need a "consume" ordering as we will generally be reading though
+        // this value (by calling the function we dlsym'ed). Rust lacks this
+        // ordering, so we have to go with the next strongest: Acquire/Release.
+        // As noted in libstd, this might be unnecessary.
+        let mut addr = self.addr.load(Ordering::Acquire);
+        if addr == Self::UNINIT {
+            let symbol = self.name.as_ptr() as *const _;
+            addr = unsafe { libc::dlsym(libc::RTLD_DEFAULT, symbol) };
+            self.addr.store(addr, Ordering::Release);
```

Comment:
Since the initialization code can run several times without impacting correctness and these load/stores are not related to any other memory operations, I think using `Relaxed` ordering should be correct.

Reply:
These load/stores relate to memory operations within `dlsym`. Usually I'd err on the side of caution per [this comment in libstd's `weak.rs`](https://github.com/rust-lang/rust/blob/1.61.0/library/std/src/sys/unix/weak.rs#L105-L121) from https://github.com/rust-lang/rust/pull/79039.

However, I may be convinced that relaxed is fine here. @thomcc, can you please weigh in?

Reply:
>These load/stores relate to memory operations within dlsym.

In this case `Release` store should be sufficient, while the load can be `Relaxed`.

Reply:
I updated the implementation here to exactly match [the implementation in `libstd`](https://github.com/rust-lang/rust/blob/282445a288b3c409a998eb72e51827537e81abd3/library/std/src/sys/unix/weak.rs#L101-L126) and I clarified some comments.

Note that we need a `Release` op and an `Acquire` op to get the "ordered before" relationship. A `Relaxed` and `Release` op don't give us that.

Reply:
Yes, this needs acquire. It’s not true that they don’t relate to other memory operations — you read through the pointer (by calling it), which counts.